RFCs/FS-1092-Erased-Union-Types.md

```diff
@@ -0,0 +1,227 @@
+```
+
+## Type inference
+[inference]: #type-inference
+
+Erased Union type is explicitly inferred meaning that at least one of the type in an expression must contain the erased union type.
+
+i.e something like the following is invalid:
+
+```fsharp
+let intOrString = if true then 1 else "Hello" // invalid
+```
+
+However the following is valid:
+
+```fsharp
+// inferred to (int|string)
+let intOrString = if true then 1 :> (int|string) else "Hello" :> _
+```
+
+This respects the rules around where explicit upcasting is required including cases despite where type information being available. Although the later might change depending on the outcome of [fslang-suggestion#849](https://github.com/fsharp/fslang-suggestions/issues/849)
+
+## Exhaustivity checking
+[exhaustivity]: #exhaustivity-checking
+
+If the selector of a pattern match is an erased union type, the match is considered exhaustive if all parts of the erased union are covered. There would be no need for fallback switch.
+
+```fsharp
+let prettyPrint (x: (int8|int16|int64|string)) =
+    match x with
+    | :? (int8|int16|int64) as y -> prettyPrintNumber y
+    | :? string as y -> prettyPrintNumber y
+```
+
+The above is the same as F# in current form:
+
+```fsharp
+let prettyPrint (x: obj) =
+    match x with
+    | :? int8 | :? int16 | :? int64 as y -> prettyPrintNumber y
+    | :? string as y -> prettyPrintNumber y
+```
+
+Similarly the following would also be considered exhaustive:
+
+```fsharp
+let prettyPrint (x: (int8|int16|int64|string)) =
+    match x with
+    | :? System.ValueType as y -> prettyPrintNumber y // int8, int16 and int64 are subtype of ValueType
+    | :? string as y -> prettyPrintNumber y
+```
+
+## Erased Type
+[erasedtype]: #erased-type
+
+The IL wrapping type for `(A | B)` is the _smallest intersection type_ of base
+types of `A` and `B`. For example:
+
+```fsharp
+// wrapping type is System.Object
+type IntOrString = (int|string)
+// wrapping type is System.ValueType
+type IntOrString = (int8|int16|float)
+type I = interface end
+type A = inherit I
+type B = inherit I
+// I is the wrapping type
+type AorB = (A|B) 
+
+type I2 = interface end
+type C = inherit I inherit I2
+type D = inherit I inherit I2
+// Both I or I2 could be potential wrapping type. The compiler would choose I2 since its the earliest ancestor
+type CorD = (C|D) 
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+TBD
+
+# Alternatives
+[alternatives]: #alternatives
+
+TBD
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+* Initial implementation should not allow for using uom in erased unions when the underlying primitive is already part of union ?
+
+    ```fsharp
+    type [<Measure>] userid
+    type UserId = int<userid>
+    type IntOrUserId = (int|UserId)
+    ```
+
+    Alternatively we could just warn when such constructs are used.
+
+* Initial implementation should not allow using static or generic type arguments in erased unions?
```

Comment:
It should definitely support generics in the first version

Reply:
@cartermp Things like 
```fsharp
let a: list<(string|int>) = ...
``` 
would work. 

But I am not sure how generics parameter in union can even be supported? As it is erased, the wrapping type cannot be known when creating IL for generics? My knowledge in this area is limited. Am I missing something?

Would you expect to write in form of 
```fsharp
let formUnion<'a, 'b> (test: 'a): ('a|'b) = test :> _
```

or 
```fsharp
let formUnion<'a, 'b when 'a :> ('a |'b) and 'b :> ('a | 'b)> (test: 'a) = test :> _
```

Reply:
I suppose the most obvious solution is using the closest common ancestor as the generic type behind the scenes. That would often be `obj`, causing boxing for any value types among the constituent types. On the other hand, given this type

```fsharp
// wrapping type is System.Object
type IntOrString = (int|string)
```

`int`s that you pass around this way will need to be boxed as well.

However, if you were to use a proper union to represent `IntOrString`, you'd still need to allocate... Perhaps boxing should be seen as an unovoidable fact?

Reply:
That would be foregoing where Erased type is `ValueType` in situations where if all cases are structs. But F# appears to do box and unbox anyway in case struct is upcasted to ValueType.

Reply:
Yeah, it does https://sharplab.io/#v2:DYLgZgzgPsCmAuACAhiRBlAnhesC2AdAGrLACusAKpgA6wDaAuogLyL1SICMA3IgEwREURkA
It makes sense when you think about it -- individual value types can have various sizes, how would you pack them in an array?

Reply:
That makes sense. Only situation I imagine where this might be bit overhead is when reduction of joins of same type or join of subtype to supertype leads to same type. For example: `(int|int)` or `(IA | I)` where `I :> IA`. But union case are only special cases though

Reply:
@cartermp I'm inclined to say that this feature should not support naked generic type variables, e.g. `(int | 'T)`.  This is just not needed to get the utility of this feature and brings in a huge range of issues which we can't solve (e.g. what happens on substitution for T).  If you want naked generic type variables you use a tagged union